### ISSUE TITLE
Entities duplicated after reboot with 0.5.2

### ISSUE THREAD

**Issue by jason0x43** [jason0x43/hacs-hubitat#51]
See [this comment](https://community.home-assistant.io/t/custom-component-hubitat/166750/495) in the forums:

> Noticed Hubitat entities intermittently get duplicated after reboot where old ones become unavailable.

**Comment by burrisd** [comment]
I've seen this a couple of times. It's late and I'm tired. Seems like the following sequence gets me back:

1. Delete all of the errored entities.
2. Correct manually the newly created entity ids to match the previous (original) that you just deleted.
3. Go to the Hubitat Hub and clear the "URL to send device events to by POST" field.
4. Reboot the Home Assistant server.

Observation: I did this procedure without clearing the POST entry and got the renamed entities again. Repeating and INCLUDING step 3 everything matched up correctly again.

Does this have anything to do with the Hubitat Hub still having the old POST set when the Home Assistant comes up to assign a new POST with a new port number?

Maybe just the intermittent you described.

**Comment by jason0x43** [comment]
I don't know why step 3 would matter off the top of my head. The integration doesn't read the POST URL from Hubitat at any point, it only sets it.

At startup, if you have [debug logging](https://github.com/jason0x43/hacs-hubitat#logging) enabled, do you see a set of messages that look like this?

```
2020-05-24 12:40:46 DEBUG (MainThread) [custom_components.hubitat.device] Initializing Hubitat hub with event server on port 0
2020-05-24 12:40:46 DEBUG (MainThread) [hubitatmaker.hub] Setting host to 10.0.1.99
2020-05-24 12:40:46 DEBUG (MainThread) [hubitatmaker.hub] Set mac to ab:cd:ef:12:34:56
2020-05-24 12:40:46 INFO (MainThread) [hubitatmaker.hub] Created hub <Hub host=10.0.1.99 app_id=2269>
2020-05-24 12:40:46 DEBUG (MainThread) [hubitatmaker.hub] Listening on 10.0.1.206:42827
2020-05-24 12:40:46 INFO (MainThread) [hubitatmaker.hub] Setting event update URL to http://10.0.1.206:42827
2020-05-24 12:40:47 DEBUG (MainThread) [hubitatmaker.hub] Loaded device list
2020-05-24 12:40:47 DEBUG (MainThread) [hubitatmaker.hub] Loading device 6
...
2020-05-24 12:40:56 DEBUG (MainThread) [hubitatmaker.hub] Loaded device 2215
2020-05-24 12:40:56 DEBUG (MainThread) [hubitatmaker.hub] Connected to Hubitat hub at 10.0.1.99
2020-05-24 12:40:56 DEBUG (MainThread) [custom_components.hubitat.device] Registered platforms
2020-05-24 12:40:56 INFO (MainThread) [custom_components.hubitat] Hubitat is ready
```

After Hubitat is initialized, you should see sets of messages that look like:

```
2020-05-24 12:40:56 DEBUG (MainThread) [custom_components.hubitat.entities] Migrating unique_ids for binary_sensor...
2020-05-24 12:40:56 DEBUG (MainThread) [custom_components.hubitat.entities] Checking for existence of entity 10.0.1.99::2269::14::acceleration...
2020-05-24 12:40:56 DEBUG (MainThread) [custom_components.hubitat.entities] Checking for existence of entity 10.0.1.99::2269::1122::acceleration...
2020-05-24 12:40:56 DEBUG (MainThread) [custom_components.hubitat.entities] Checking for existence of entity 10.0.1.99::2269::1954::acceleration...
2020-05-24 12:40:56 DEBUG (MainThread) [custom_components.hubitat.entities] Added HubitatAccelerationSensor entities: [<Entity Barn Sensor acceleration: off>, <Entity Garage Sensor acceleration: off>, <Entity Breezeway Sensor acceleration: off>]
```

If some entities still need migration, you'll also see messages about specific entities being migrated.

My best guess about the duplicate IDs at this point is that entity migration (which updates the internal unique IDs of entities from the pre-v0.5.0 format to the v0.5.0 format) isn't working properly for some reason. The old entities are sticking around, and when you restart HA, new ones are being created again.

**Comment by burrisd** [comment]
Thanks for the prompt response!

I'm still a bit new to both Hubitat and HA. Let's say the learning curve is a bit steep! :)

I wasn't sure that this made any sense, but after repeating it a couple of times I thought I would mention it in case it rang any bells.

I was able to do some flows to do much of my initial stuff until I was able to figure out how to get the integration successfully loaded. There's a lot of info out there but not much tying it all together. After getting the integration working I'm now wrestling with the various "personalities" of nodes and the Maker API.

One thing making matters worse is that I was trying to test a flow and the behavior you describe occurred multiple times along with the fact that the switch I am testing with does not send out events until the light is actually on. If I toggle too fast, no messages are sent at all. Since I am coincidentally testing that very thing (duration of a switch toggle) it was all quite confusing.

I'll let you know if I discover anything I think might be useful. I'm debating how far into the internals of building nodes and integrations I want to get.

**Comment by jason0x43** [comment]
> Let's say the learning curve is a bit steep! :)

It is so very, very steep. 😄

>  some flows

What do you mean by a "flow"?

> the switch I am testing with does not send out events until the light is actually on

Are you listening to HA or the Maker API? HA should send out a state update as soon as you toggle something in the UI. When you flip a toggle in the HA UI, the associated HA entity updates its state immediately. HA will wait for a short time then re-check the actual device state to see if it changed; if it didn't, HA will reset the entity's state to what it was before you flipped the toggle.

**Comment by burrisd** [comment]
Sorry, Node-Red terminology. (Nodes, flows, etc.)

I'm talking about the physical Z-Wave switch. Hubitat is not seeing the event until the light on the switch comes on. There's a delay between the toggle and the activation. I assume it's software debouncing. I was caught off guard at the switch not sending events if the switch toggles too quickly. Naturally if the switch doesn't generate an event, HA is not going to know. 

There also seemed to be a few quirks between Z-Wave switches and Zigbee bulbs that I have yet to define. I was intermittently noticing subtle differences in behavior that I would be hard-pressed to describe because I was FOCUSED on WTF was going on with my experiment.

I actually did notice that HA behavior you're describing. Interestingly enough that was adding to the confusion enough to eliminate that in lieu of toggling the actual switch. Once I've got it figured out I can work around the issues as needed. Naturally if HA is going to autonomously generate its own events, reliance on switch timing for controlling a "hidden" feature might get tricky.

**Comment by alaghbar** [comment]
Just got another instance of entities duplication after properly shutting down both hassio (Rp) and Hubitat - had to replace UPS.

Followed your instruction and found the file "core.entity_registry". Here's a sample duplicated entity:

{
                "capabilities": {},
                "config_entry_id": "ff0b6184abee45e1aacc9425a3aacf0d",
                "device_class": "garage_door",
                "device_id": "65f03b68ee7047ae871e71cefa388d9a",
                "disabled_by": null,
                "entity_id": "binary_sensor.garage_door_contact",
                "icon": null,
                "name": null,
                "original_icon": null,
                "original_name": "Garage Door contact",
                "platform": "hubitat",
                "supported_features": 0,
                "unique_id": "34:e1:d1:80:51:fa::834::8::contact",
                "unit_of_measurement": null
            },
{
                "capabilities": null,
                "config_entry_id": "ff0b6184abee45e1aacc9425a3aacf0d",
                "device_class": "garage_door",
                "device_id": "65f03b68ee7047ae871e71cefa388d9a",
                "disabled_by": null,
                "entity_id": "binary_sensor.garage_door_contact_2",
                "icon": null,
                "name": null,
                "original_icon": null,
                "original_name": "Garage Door contact",
                "platform": "hubitat",
                "supported_features": 0,
                "unique_id": "00:00:00:00:00:00::834::8::contact",
                "unit_of_measurement": null
            }

Please let me if you require any further detail. Thank you

**Comment by jason0x43** [comment]
It looks like the duplicate is created because the integration was unable to determine the MAC address of your hub (the `00:00:00:00:00:00`). Are you using `localhost` or `127.0.0.1` for the Hubitat IP address, or are you accessing it through a proxy?

**Comment by alaghbar** [comment]
Interesting!!!

No. I'm using the proper IP address of Habitat Hub:

![IMG_0025](https://user-images.githubusercontent.com/59179396/82846293-05fe2680-9eb6-11ea-809f-ad944876e567.jpg)

binary_sensor.car_door_opener_contact has status "Unavailable"
binary_sensor.car_door_opener_contact_2 has the proper and active status, very strange!!

![IMG_0026](https://user-images.githubusercontent.com/59179396/82846354-6ab98100-9eb6-11ea-9a4c-5a29387677c4.jpg)

**Comment by alaghbar** [comment]
One more note, not sure how relevant. Just checked Maker API settings:
![HASS](https://user-images.githubusercontent.com/59179396/82846539-53c75e80-9eb7-11ea-99be-9f6c46532663.png)
I did not enter this.... IP of HASS.

**Comment by burrisd** [comment]
The UI is retaining the association you set up that now point to an invalid id. This actually turned out good because it allowed me to correct using the steps I outlined earlier.

I've turned on the debugging and, of course, the reboot was fine. No duplicates at this point. I am seeing the messages you mentioned in the log. I'll forward results if I see the problem occur again.

Seem like from the debug above, the config_entry_id and the device_id matching is a clue that perhaps another check is required to verify before creating the duplicate, or at least to confirm deletion of the old one in the process. These values all must be unique in the configuration, correct?

**Comment by burrisd** [comment]
Ok, reboot of HA did not show problem. However, reboot of shutdown of HA, reboot Hubitat, start HA, did exhibit the problem. Note that this could still be the intermittent issue you described.

![image](https://user-images.githubusercontent.com/65719766/82848296-8cb40300-9eb8-11ea-96ba-67935353f4d6.png)
![image](https://user-images.githubusercontent.com/65719766/82848322-a6ede100-9eb8-11ea-8a9c-6a975330ca76.png)
![image](https://user-images.githubusercontent.com/65719766/82848348-ba00b100-9eb8-11ea-8163-8538abaa69bb.png)

2020-05-25 18:40:16 WARNING (MainThread) [homeassistant.components.climate] ClimateDevice is deprecated, modify HubitatThermostat to extend ClimateEntity
2020-05-25 18:40:16 INFO (MainThread) [homeassistant.components.climate] Setting up climate.hubitat
2020-05-25 18:40:16 WARNING (MainThread) [homeassistant.components.lock] LockDevice is deprecated, modify HubitatLock to extend LockEntity
2020-05-25 18:40:16 INFO (MainThread) [homeassistant.components.lock] Setting up lock.hubitat
2020-05-25 18:40:16 WARNING (MainThread) [homeassistant.components.cover] CoverDevice is deprecated, modify HubitatCover to extend CoverEntity
2020-05-25 18:40:16 WARNING (MainThread) [homeassistant.components.cover] CoverDevice is deprecated, modify HubitatDoorControl to extend CoverEntity
2020-05-25 18:40:16 WARNING (MainThread) [homeassistant.components.cover] CoverDevice is deprecated, modify HubitatGarageDoorControl to extend CoverEntity
2020-05-25 18:40:16 WARNING (MainThread) [homeassistant.components.cover] CoverDevice is deprecated, modify HubitatWindowShade to extend CoverEntity
2020-05-25 18:40:16 WARNING (MainThread) [homeassistant.components.light] Light is deprecated, modify HubitatLight to extend LightEntity
2020-05-25 18:40:16 WARNING (MainThread) [homeassistant.components.switch] SwitchDevice is deprecated, modify HubitatSwitch to extend SwitchEntity
2020-05-25 18:40:16 WARNING (MainThread) [homeassistant.components.switch] SwitchDevice is deprecated, modify HubitatPowerMeterSwitch to extend SwitchEntity
2020-05-25 18:40:16 WARNING (MainThread) [homeassistant.components.switch] SwitchDevice is deprecated, modify HubitatAlarm to extend SwitchEntity
2020-05-25 18:40:16 INFO (MainThread) [homeassistant.components.switch] Setting up switch.hubitat
2020-05-25 18:40:16 INFO (MainThread) [homeassistant.components.fan] Setting up fan.hubitat
2020-05-25 18:40:16 INFO (MainThread) [homeassistant.components.light] Setting up light.hubitat
2020-05-25 18:40:16 WARNING (MainThread) [homeassistant.components.alarm_control_panel] AlarmControlPanel is deprecated, modify HubitatSecurityKeypad to extend AlarmControlPanelEntity
2020-05-25 18:40:16 INFO (MainThread) [homeassistant.components.alarm_control_panel] Setting up alarm_control_panel.hubitat
2020-05-25 18:40:16 INFO (MainThread) [homeassistant.components.cover] Setting up cover.hubitat
2020-05-25 18:40:16 DEBUG (MainThread) [custom_components.hubitat.entities] Migrating unique_ids for lock...
2020-05-25 18:40:16 DEBUG (MainThread) [custom_components.hubitat.entities] Checking for existence of entity 10.10.10.30::133::8...
2020-05-25 18:40:16 DEBUG (MainThread) [custom_components.hubitat.entities] Added HubitatLock entities: [<Entity Front door lock: locked>]
2020-05-25 18:40:16 DEBUG (MainThread) [custom_components.hubitat.entities] Migrating unique_ids for switch...
2020-05-25 18:40:16 DEBUG (MainThread) [custom_components.hubitat.entities] Checking for existence of entity 10.10.10.30::133::5...
2020-05-25 18:40:16 DEBUG (MainThread) [custom_components.hubitat.entities] Checking for existence of entity 10.10.10.30::133::33...
2020-05-25 18:40:16 DEBUG (MainThread) [custom_components.hubitat.entities] Checking for existence of entity 10.10.10.30::133::34...
2020-05-25 18:40:16 DEBUG (MainThread) [custom_components.hubitat.entities] Checking for existence of entity 10.10.10.30::133::35...
2020-05-25 18:40:16 DEBUG (MainThread) [custom_components.hubitat.entities] Checking for existence of entity 10.10.10.30::133::169...
2020-05-25 18:40:16 DEBUG (MainThread) [custom_components.hubitat.entities] Added HubitatSwitch entities: [<Entity Kitchen accent outlet: on>, <Entity Closet LED: on>, <Entity Closet air: off>, <Entity Closet Water: off>, <Entity Bedtime switch: off>]
/config/custom_components/hubitat/switch.py:144: RuntimeWarning: coroutine 'create_and_add_event_emitters' was never awaited
  create_and_add_event_emitters(hass, entry, is_button_controller)
RuntimeWarning: Enable tracemalloc to get the object allocation traceback
2020-05-25 18:40:16 DEBUG (MainThread) [custom_components.hubitat.entities] Migrating unique_ids for light...
2020-05-25 18:40:16 DEBUG (MainThread) [custom_components.hubitat.entities] Checking for existence of entity 10.10.10.30::133::1...
2020-05-25 18:40:16 DEBUG (MainThread) [custom_components.hubitat.entities] Checking for existence of entity 10.10.10.30::133::2...
2020-05-25 18:40:16 DEBUG (MainThread) [custom_components.hubitat.entities] Checking for existence of entity 10.10.10.30::133::3...
2020-05-25 18:40:16 DEBUG (MainThread) [custom_components.hubitat.entities] Checking for existence of entity 10.10.10.30::133::4...
2020-05-25 18:40:16 DEBUG (MainThread) [custom_components.hubitat.entities] Checking for existence of entity 10.10.10.30::133::65...
2020-05-25 18:40:16 DEBUG (MainThread) [custom_components.hubitat.entities] Checking for existence of entity 10.10.10.30::133::98...
2020-05-25 18:40:16 DEBUG (MainThread) [custom_components.hubitat.entities] Checking for existence of entity 10.10.10.30::133::99...
2020-05-25 18:40:16 DEBUG (MainThread) [custom_components.hubitat.entities] Checking for existence of entity 10.10.10.30::133::100...
2020-05-25 18:40:16 DEBUG (MainThread) [custom_components.hubitat.entities] Added HubitatLight entities: [<Entity Lower hallway lights: off>, <Entity Front lights: off>, <Entity Upper hallway lights: off>, <Entity Stairway lights: off>, <Entity Bedroom lamp: on>, <Entity Deck light: off>, <Entity Living room lamp: off>, <Entity Loft lamp: on>]
2020-05-25 18:40:16 INFO (MainThread) [homeassistant.helpers.entity_registry] Registered new lock.hubitat entity: lock.front_door_lock_2
2020-05-25 18:40:16 INFO (MainThread) [homeassistant.helpers.entity_registry] Registered new switch.hubitat entity: switch.kitchen_accent_outlet_2
2020-05-25 18:40:16 INFO (MainThread) [homeassistant.helpers.entity_registry] Registered new switch.hubitat entity: switch.closet_led_2
2020-05-25 18:40:16 INFO (MainThread) [homeassistant.helpers.entity_registry] Registered new switch.hubitat entity: switch.closet_air_2
2020-05-25 18:40:16 INFO (MainThread) [homeassistant.helpers.entity_registry] Registered new switch.hubitat entity: switch.closet_water_2
2020-05-25 18:40:16 INFO (MainThread) [homeassistant.helpers.entity_registry] Registered new switch.hubitat entity: switch.bedtime_switch_2
2020-05-25 18:40:16 INFO (MainThread) [homeassistant.helpers.entity_registry] Registered new light.hubitat entity: light.lower_hallway_lights_2
2020-05-25 18:40:16 INFO (MainThread) [homeassistant.helpers.entity_registry] Registered new light.hubitat entity: light.front_lights_2
2020-05-25 18:40:16 INFO (MainThread) [homeassistant.helpers.entity_registry] Registered new light.hubitat entity: light.upper_hallway_lights_2
2020-05-25 18:40:16 INFO (MainThread) [homeassistant.helpers.entity_registry] Registered new light.hubitat entity: light.stairway_lights_2
2020-05-25 18:40:16 INFO (MainThread) [homeassistant.helpers.entity_registry] Registered new light.hubitat entity: light.bedroom_lamp_2
2020-05-25 18:40:16 INFO (MainThread) [homeassistant.helpers.entity_registry] Registered new light.hubitat entity: light.deck_light_2
2020-05-25 18:40:16 INFO (MainThread) [homeassistant.helpers.entity_registry] Registered new light.hubitat entity: light.living_room_lamp_2
2020-05-25 18:40:16 INFO (MainThread) [homeassistant.helpers.entity_registry] Registered new light.hubitat entity: light.loft_lamp_2
2020-05-25 18:35:50 INFO (MainThread) [homeassistant.components.sensor] Setting up sensor.upnp

**Comment by burrisd** [comment]
Additional info:

After seeing the report above from alaghbar it seemed that the issue was getting MAC address. This got me thinking that my testing setup was a bit unusual in that I was using a USB Wifi adapter to join my desktop computer to the Wifi network. I hadn't given this much though but then it occurred to me that I also had DHCP issues with the VirtualBox VM because the VM was reporting the same MAC address as the adapter.

Apparently there is a difference in resolving MAC address when sharing a WLAN.

I have moved my attention back to the Pi now that I have apparently figured out my issues with getting this working on the Pi. I have now imported all my Node Red JSON and am reconstructing my setup on the Pi with a more typical setup and the Pi using its own LAN interface. So far, so good.

I should know soon if the problem still occurs.
